src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.PackTool.targets

```diff
@@ -68,30 +67,106 @@ Copyright (c) .NET Foundation. All rights reserved.
+    <_ShimInputCacheFile>$([MSBuild]::NormalizePath($(MSBuildProjectDirectory), $(_ShimInputCacheFile)))</_ShimInputCacheFile>
+    <_ShimCreatedSentinelFile Condition="'$(_ShimCreatedSentinelFile)' == ''">$(IntermediateOutputPath)$(MSBuildProjectName).shimcreated.sentinel</_ShimCreatedSentinelFile>
+    <_ShimCreatedSentinelFile>$([MSBuild]::NormalizePath($(MSBuildProjectDirectory), $(_ShimCreatedSentinelFile)))</_ShimCreatedSentinelFile>
+  </PropertyGroup>
+
+  <Target Name="GenerateShimsAssets"
+          BeforeTargets="CopyFilesToOutputDirectory"
+          DependsOnTargets="ResolvePackageAssets;_PackToolValidation;_GenerateShimInputCache;_ComputeExpectedEmbeddedApphostPaths"
+          Condition="'$(PackAsTool)' == 'true'"
+          Inputs="@(_GenerateShimsAssetsInput)"
+          Outputs="$(_ShimCreatedSentinelFile)">
 
     <PropertyGroup>
-      <PackagedShimOutputRootDirectory Condition=" '$(PackagedShimOutputRootDirectory)' == '' ">$(IntermediateOutputPath)</PackagedShimOutputRootDirectory>
+      <PackagedShimOutputRootDirectory Condition=" '$(PackagedShimOutputRootDirectory)' == '' ">$(OutDir)</PackagedShimOutputRootDirectory>
     </PropertyGroup>
 
     <GenerateShims
-      DotNetAppHostExecutableNameWithoutExtension="$(_DotNetAppHostExecutableNameWithoutExtension)"
+      ApphostsForShimRuntimeIdentifiers="@(_ApphostsForShimRuntimeIdentifiers)"
       PackagedShimOutputDirectory="$(PackagedShimOutputRootDirectory)/shims/$(TargetFramework)"
       PackageId="$(PackageId)"
       PackageVersion="$(Version)"
-      ProjectAssetsFile="$(ProjectAssetsFile)"
-      ProjectPath="$(MSBuildProjectFullPath)"
       ShimRuntimeIdentifiers="@(_PackAsToolShimRuntimeIdentifiers)"
       TargetFrameworkMoniker="$(NuGetTargetMoniker)"
       ToolCommandName="$(ToolCommandName)"
-      ToolEntryPoint="$(ToolEntryPoint)"
-      >
+      ToolEntryPoint="$(ToolEntryPoint)">
 
       <Output TaskParameter="EmbeddedApphostPaths" ItemName="_EmbeddedApphostPaths" />
     </GenerateShims>
+
+    <!-- Workaround for https://github.com/dotnet/corefx/issues/31379
+         LastWriteTime in shims are not accurate. And _ShimInputCacheFile will have later timestamp than generated shims.
+         Use a created file to "record" LastWriteTime. And only use it in "Outputs" field for timestamp comparison.
+    -->
+    <WriteLinesToFile Lines="This file's LastWriteTime is used in incremental build" File="$(_ShimCreatedSentinelFile)" Overwrite="True" />
+
+    <ItemGroup>
+      <!-- Do this in an ItemGroup instead of as an output parameter so that it still gets added to the item set
+           during incremental builds when the task is skipped -->
+      <FileWrites Include="@(_EmbeddedApphostPaths)" Condition="Exists('@(_EmbeddedApphostPaths)')"/>
+      <FileWrites Include="$(_ShimCreatedSentinelFile)" Condition="Exists('$(_ShimCreatedSentinelFile)')"/>
```

Comment:
I didn't know that Exists works on items. What does it return if one exists, but another doesn't.

I don't think it's necessary to condition adds to filewrites

Reply:
removed condition